pkg/policy/mapstate.go

```diff
@@ -498,8 +526,178 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
 				}
 			}
 		}
-		keys.redirectPreferredInsert(newKey, newEntry, adds, deletes)
+		keys.authPreferredInsert(newKey, newEntry, features, adds, deletes)
+	}
+}
+
+// IsSuperSetOf checks if the receiver Key is a superset of the argument Key, and returns a
+// specificity score of the receiver key (higher score is more specific), if so. Being a superset
+// means that the receiver key would match all the traffic of the argument key without being the
+// same key. Hence, a L3-only key is not a superset of a L4-only key, as the L3-only key would match
+// the traffic for the given L3 only, while the L4-only key matches traffic on the given port for
+// all the L3's.
+// Returns 0 if the receiver key is not a superset of the argument key.
+//
+// Specificity score for all possible superset wildcard patterns. Datapath requires proto to be specified if port is specified.
+// x. L3/proto/port
+//  1. */*/*
+//  2. */proto/*
+//  3. */proto/port
+//  4. ID/*/*
+//  5. ID/proto/*
+//     ( ID/proto/port can not be superset of anything )
+func (k Key) IsSuperSetOf(other Key) int {
+	if k.TrafficDirection != other.TrafficDirection {
+		return 0 // TrafficDirection must match for 'k' to be a superset of 'other'
+	}
+	if k.Identity == 0 {
+		if other.Identity == 0 {
+			if k.Nexthdr == 0 { // k.DestPort == 0 is implied
+				if other.Nexthdr != 0 {
+					return 1 // */*/* is a superset of */proto/x
+				} // else both are */*/*
+			} else if k.Nexthdr == other.Nexthdr {
+				if k.DestPort == 0 && other.DestPort != 0 {
+					return 2 // */proto/* is a superset of */proto/port
+				} // else more specific or different ports
+			} // else more specific or different protocol
+		} else {
+			// Wildcard L3 is a superset of a specific L3 only if wildcard L3 is also wildcard L4, or the L4's match between the keys
+			if k.Nexthdr == 0 { // k.DestPort == 0 is implied
+				return 1 // */*/* is a superset of ID/x/x
+			} else if k.Nexthdr == other.Nexthdr {
+				if k.DestPort == 0 {
+					return 2 // */proto/* is a superset of ID/proto/x
+				} else if k.DestPort == other.DestPort {
+					return 3 // */proto/port is a superset of ID/proto/port
+				} // else more specific or different ports
+			} // else more specific or different protocol
+		}
+	} else if k.Identity == other.Identity {
+		if k.Nexthdr == 0 {
+			if other.Nexthdr != 0 {
+				return 4 // ID/*/* is a superset of ID/proto/x
+			} // else both are ID/*/*
+		} else if k.Nexthdr == other.Nexthdr {
+			if k.DestPort == 0 && other.DestPort != 0 {
+				return 5 // ID/proto/* is a superset of ID/proto/port
+			} // else more specific or different ports
+		} // else more specific or different protocol
+	} // else more specific or different identity
+	return 0
+}
+
+// authPreferredInsert applies AuthType of a more generic entry to more specific entries, if not
+// explicitly specified.
+func (keys MapState) authPreferredInsert(newKey Key, newEntry MapStateEntry, features policyFeatures, adds, deletes Keys) {
```

Comment:
This function and the logic in `IsSuperSetOf()` seems awfully similar to equivalent logic in deny policy logic handling. On one hand, having separate implementations seems like potentially a good way to prevent auth policies from causing a regression in deny policies. On the other hand, I'm concerned that now we may end up with two entirely different implementations of conflict resolution for map entries.

Is there a way to share the core implementation with deny preferred inserts here? I'd imagine there are maybe some subtle differences in the implementations, but I feel like supporting one giant set of confusing conflict resolution statements that generically applies to both cases would be easier than maintaining two entirely different sets.

Reply:
I've reached my policy confusion quota for the day, so I've stopped before reviewing any of this prefered insert logic.

Reply:
Keeping them separate makes it obvious that auth policies can not revert any deny rules with non-deny rules. The logic is also subtly different and I'd rather get it right first and optimise by combining the loops later.

Reply:
Ack, reasonable tradeoff. Is there a TL;DR/explanation on the subtly different logic?